Detectors/TPC/workflow/src/CATrackerSpec.cxx

```diff
@@ -644,34 +496,26 @@ DataProcessorSpec getCATrackerSpec(ca::Config const& specconfig, std::vector<int
     if (specconfig.caClusterer) {
       // We accept digits and MC labels also if we run on ZS Raw data, since they are needed for MC label propagation
       if (!specconfig.zsDecoder) { // FIXME: We can have digits input in zs decoder mode for MC labels, to be made optional
-        inputs.emplace_back(InputSpec{"input", gDataOriginTPC, "DIGITS", 0, Lifetime::Timeframe});
+        inputs.emplace_back(InputSpec{"input", ConcreteDataTypeMatcher{gDataOriginTPC, "DIGITS"}, Lifetime::Timeframe});
       }
     } else {
-      inputs.emplace_back(InputSpec{"input", gDataOriginTPC, "CLUSTERNATIVE", 0, Lifetime::Timeframe});
+      inputs.emplace_back(InputSpec{"input", ConcreteDataTypeMatcher{gDataOriginTPC, "CLUSTERNATIVE"}, Lifetime::Timeframe});
     }
     if (specconfig.processMC) {
       if (specconfig.caClusterer) {
-        constexpr o2::header::DataDescription datadesc("DIGITSMCTR");
         if (!specconfig.zsDecoder) { // FIXME: We can have digits input in zs decoder mode for MC labels, to be made optional
-          inputs.emplace_back(InputSpec{"mclblin", gDataOriginTPC, datadesc, 0, Lifetime::Timeframe});
+          inputs.emplace_back(InputSpec{"mclblin", ConcreteDataTypeMatcher{gDataOriginTPC, "DIGITSMCTR"}, Lifetime::Timeframe});
         }
       } else {
-        inputs.emplace_back(InputSpec{"mclblin", gDataOriginTPC, "CLNATIVEMCLBL", 0, Lifetime::Timeframe});
+        inputs.emplace_back(InputSpec{"mclblin", ConcreteDataTypeMatcher{gDataOriginTPC, "CLNATIVEMCLBL"}, Lifetime::Timeframe});
       }
     }
 
-    auto tmp = std::move(mergeInputs(inputs, tpcsectors.size(),
-                                     [tpcsectors](InputSpec& input, size_t index) {
-                                       // using unique input names for the moment but want to find
-                                       // an input-multiplicity-agnostic way of processing
-                                       input.binding += std::to_string(tpcsectors[index]);
-                                       DataSpecUtils::updateMatchingSubspec(input, tpcsectors[index]);
-                                     }));
     if (specconfig.zsDecoder) {
       // We add this after the mergeInputs, since we need to keep the subspecification
-      tmp.emplace_back(InputSpec{"zsraw", ConcreteDataTypeMatcher{"TPC", "RAWDATA"}, Lifetime::Timeframe});
+      inputs.emplace_back(InputSpec{"zsraw", ConcreteDataTypeMatcher{"TPC", "RAWDATA"}, Lifetime::Timeframe});
     }
-    return tmp;
+    return inputs;
   };
 
```

Comment:
@matthiasrichter @shahor02 @ktf : I have a question here, which I hope you might be able to clarify:
I have removed that mergeInput function which duplicates all the inputs explicitly adding one per TPC sector with corresponding subspec, and instead I I use the `ConcreteDataTypeMatcher` with a catchall for the subspec.

The reason for this is that I want to have 36 messages with subspec 0 to 35 in case we get the input as one per sector, and a single message with subspec 37 if there is a single linear input.

I need that the processing function (Line 292: `auto processingFct = [processAttributes, specconfig](ProcessingContext& pc) {`) is called only once with all inputs.
This is working for the `CLUSTERNATIVE` input type (when I use the old clusterer, or read the clusters from a file), but when I use the `DIGITS` input type, it doesn't work, and the processing function is called once per subspecification, with only that subspecification as input.

Without this change removing the `mergeInputs`, also the `DIGITS` input type works, but I do not understand what is the difference to `CLUSTERNATIVE`, and in general I don't know when the processing function will wait for which inputs, and when not.

Reply:
afaik, with the CATrackerSpec completion policy disabled (https://github.com/AliceO2Group/AliceO2/blob/5f973cd2c0ce2ca59d545e3309cf056e9c907423/Detectors/TPC/workflow/src/tpc-reco-workflow.cxx#L83) the processing input must be called only once, if for some input it is called multiple times, this may mean that different subspecs of this input were sent with different timestamps. But better if @matthiasrichter or @ktf comment.

Using the same ``"CLUSTERNATIVE"`` data description for different data types might be error prone, since you are anyway processing 2 cases separately, do you really need this?

Reply:
@shahor02 : My problem is: If I use different data types, every consumer of TPC clusters (tpc tracking itself, tpc its matching, tpc scd interpolation, qc) needs to register either CLUSTERNATIVE or CLUSTERNATIVEFLAT as input, so I need options for all these data types depending on whether the input comes from a root file / old clusterizer / new clusterizer.

I thought we discussed above that we want to avoid this.

As I do it now, it is totally transparent to the consumers. If it comes from the ca-clusterer, it comes as a single buffer, otherwise I get 36 messages with different subspec. This is already fully working. The only thing I must make sure is that I get always all messages at once, which means either the 36 messages with subspec 0-35, or the single message with subspec 36. How can I achieve this?

I have found out that the different behavior between `CLUSTERNATIVE` and `DIGITS` comes from the tpc cluster writer. I think when the TPC cluster writer is active, it registers separate inputs for all `CLUSTERNATIVE`, forcing them to be synchronized, and then other consumers get them together as well. When I pass the `disable-writer` option, I also get individual messages for `CLUSTERNATIVE`, as I do for `DIGITS`.

Just so I understand correctly: This behavior is expected, and it is because the callback to `getCATrackerCompletionPolicy` is currently commented out? I believe you commented this out because it triggered some other problem?

Reply:
@davidrohr I wanted to limit the number of formats, not the labels... Anyway, if we end up with a common helper which every consumer of CLUSTERNATIVE or CLUSTERNATIVEFLAT can use it should be ok to use the same label.
I had to comment out getCATrackerCompletionPolicy exactly because it was triggering a call of processingFct whenever some (partial) input is available (while the code of processingFct was assuming having complete set of inputs in one shot). So, to me the behaviour you see is NOT expected, unless the digits of different sectors may be sent with different time stamps (i.e. as coming from different timeframes). Here @matthiasrichter should comment